### ISSUE TITLE
[FEATURE] Support v2 volume without hugepage

### ISSUE THREAD

**Issue by derekbit** [longhorn/longhorn#7066]
## Is your feature request related to a problem? Please describe (👍 if you like this request)

hugepage is evitable for supporting v2 volume feature. Although huge page can improve the IO performance, but there are some drawbacks

- not easy to modify the size dynamically 
- dedicated memory size (waste)
- memory pressure for low-spec machines

There is a discussion in https://github.com/spdk/spdk/issues/2973 for the no hugepage support.  Although it is not implemented yet, but we can continue investigating the design, implementing it and integrating it into v2 volume.

**Comment by derekbit** [comment]
Assigned to @c3y1huang. She will dive into the spdk internals.
cc @innobead

**Comment by derekbit** [comment]
Applied the PRs https://github.com/longhorn/longhorn-manager/pull/4045 and https://github.com/longhorn/longhorn-instance-manager/pull/1069 and enabling v2 data engine, 

The two values, VmPeak and VmSize, are 2188696 kB if we set the data-engine-memory-size to 2048 MiB. 

Note: VmPeak Peak is virtual memory size, and VmSize is virtual memory size.

```
instance-manager-be923d729f85982832a698b93e0c0491:/ # ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0   2680  1132 ?        Ss   04:35   0:00 /tini -- instance-manager --spdk-log all --spdk-cpumask 0x1 --spdk-memory-size 2048 --enable-spdk --debug daemon --spdk-enabled --listen 0.0.0.0:8500 --spdk-no-h
root         7  0.0  0.0   4272  3620 ?        S    04:35   0:00 /bin/bash /usr/local/bin/instance-manager --spdk-log all --spdk-cpumask 0x1 --spdk-memory-size 2048 --enable-spdk --debug daemon --spdk-enabled --listen 0.0.0.0:
root        17 99.3  3.6 2188696 282732 ?      RLl  04:35   1:40 spdk_tgt -L all -m 0x1 --mem-size 2048 --no-huge
root        18  0.0  0.0   2696  1288 ?        S    04:35   0:00 tee -a /log/spdk_tgt.log
root        23  0.0  0.0   4272  1704 ?        S    04:35   0:00 /bin/bash /usr/local/bin/exec-logrotate
root        25  0.1  0.6 1866448 54452 ?       Sl   04:35   0:00 longhorn-instance-manager --debug daemon --spdk-enabled --listen 0.0.0.0:8500
root        26  0.0  0.0   2696  1232 ?        S    04:35   0:00 tee -a /log/instance-manager-v2.log
root        32  0.0  0.0   2692  1228 ?        S    04:35   0:00 sleep 600
root       113  0.0  0.0   5176  4600 pts/1    Ss   04:36   0:00 bash
root       236  0.0  0.0   7744  3396 pts/1    R+   04:37   0:00 ps aux


instance-manager-be923d729f85982832a698b93e0c0491:/ # cat /proc/17/status
Name:	reactor_0
Umask:	0022
State:	R (running)
Tgid:	17
Ngid:	0
Pid:	17
PPid:	7
TracerPid:	0
Uid:	0	0	0	0
Gid:	0	0	0	0
FDSize:	256
Groups:	0
NStgid:	17
NSpid:	17
NSpgid:	7
NSsid:	1
VmPeak:	 2188696 kB
VmSize:	 2188696 kB
VmLck:	    3348 kB
VmPin:	       0 kB
VmHWM:	  282732 kB
VmRSS:	  282732 kB
RssAnon:	   43104 kB
RssFile:	   10408 kB
RssShmem:	  229220 kB
VmData:	   63344 kB
VmStk:	     132 kB
VmExe:	   10180 kB
VmLib:	    9832 kB
VmPTE:	     688 kB
VmSwap:	       0 kB
HugetlbPages:	       0 kB
CoreDumping:	0
THP_enabled:	1
Threads:	2
...
```

**Comment by chriscchien** [comment]
Verified **pass** on longhorn master(https://github.com/longhorn/longhorn-instance-manager/commit/bb3bd2ca2ef7e6e48e57d43e96865a37ddd60510), v1.10.x(https://github.com/longhorn/longhorn-instance-manager/commit/287c5f0e21e90ca209d1d6edf65ed728cc955301) with test [steps](https://github.com/longhorn/longhorn/issues/7066#issuecomment-3257047512)

**Comment by shuo-wu** [comment]
I am still confused about the final solution:
1. The related option `--no-huge` is not mentioned in the official doc https://spdk.io/doc/app_overview.html at all. I am not sure what the correct way to use it is
2. Then it leads to the next question. Why do we need to set option `--mem-size`? IIUC, this means that we may still need to reserve memory for SPDK based on the explanation in the above doc. And the doc also mentioned that `Starting with DPDK 18.05.1, it's possible to reserve hugepages at runtime, meaning that SPDK application can be started with 0 pre-reserved memory.` Can we use option `--no-huge` only?
3. I don't know how much it will degrade the performance. If it does not hurt the performance too much, can we enable `--no-huge` by default? Can it benefit the v2 live upgrade workflow simplification?

**Comment by PhanLe1010** [comment]
> The related option --no-huge is not mentioned in the official doc https://spdk.io/doc/app_overview.html at all. I am not sure what the correct way to use it is

Looks like doc has not updated. I think specifying `--no-huge` is correct way to disable hugepage usage. Ref SPDK PR https://review.spdk.io/c/spdk/spdk/+/17483 

> Then it leads to the next question. Why do we need to set option --mem-size? IIUC, this means that we may still need to reserve memory for SPDK based on the explanation in the above doc. And the doc also mentioned that Starting with DPDK 18.05.1, it's possible to reserve hugepages at runtime, meaning that SPDK application can be started with 0 pre-reserved memory. Can we use option --no-huge only?

When disabling hugepage by specifying `--no-huge`, I think the flag `--mem-size` is redundant and will not be used by SPDK. I agree that we should remove it when `--no-huge` is used.  WDYT @derekbit @c3y1huang 


> I don't know how much it will degrade the performance. If it does not hurt the performance too much, can we enable --no-huge by default? Can it benefit the v2 live upgrade workflow simplification?

We can do more testing but I think it should affect performance. SPDK use huge page by default so maybe it has reason

**Comment by derekbit** [comment]
> I am still confused about the final solution:
> 
> 1. The related option `--no-huge` is not mentioned in the official doc [spdk.io/doc/app_overview.html](https://spdk.io/doc/app_overview.html) at all. I am not sure what the correct way to use it is

Yes, the doc is outdated. We can see the option from the output of `spdk_tgt --help` and the related ticket and PRs.
https://github.com/spdk/spdk/issues/2973


> 3. Then it leads to the next question. Why do we need to set option `--mem-size`? IIUC, this means that we may still need to reserve memory for SPDK based on the explanation in the above doc. And the doc also mentioned that `Starting with DPDK 18.05.1, it's possible to reserve hugepages at runtime, meaning that SPDK application can be started with 0 pre-reserved memory.` Can we use option `--no-huge` only?

When we use hugepage, we need to reserve the physical memory that is unable to share with other process.
Instead, with "--no-huge" and "--mem-size", virtual memory rather than physical memory is still required.
You can see the VmSize in test result. https://github.com/longhorn/longhorn/issues/7066#issuecomment-3257046888

> Can we use option `--no-huge` only

No, "--mem-size" is required.

```
# spdk_tgt --no-huge
Disabling hugepages requires specifying how much memory will be allocated using -s parameter
[2025-09-09 00:22:49.138067] init.c: 753:spdk_env_init: *ERROR*: Invalid arguments to initialize DPDK
[2025-09-09 00:22:49.138164] app.c: 512:app_setup_env: *ERROR*: Unable to initialize SPDK env
```
Ref: https://github.com/spdk/spdk/blob/master/lib/env_dpdk/init.c#L394-L421

> 5. I don't know how much it will degrade the performance. If it does not hurt the performance too much, can we enable `--no-huge` by default? 

Will benchmark the performance later.

> Can it benefit the v2 live upgrade workflow simplification?

Probably, yes.

**Comment by shuo-wu** [comment]
Got it. The VmSize in the process status is more like the result of setting up `--mem-size 2048`.

Anyway, we always need to reserve some memory for the v2 instance manager pods. BTW, should we align spdk `--mem-size` with the instance manager pod memory request? Although we have not introduced the corresponding setting in Longhorn.

**Comment by derekbit** [comment]
> should we align spdk --mem-size with the instance manager pod memory request

Good idea. I think we should. We can handle it when working on the resource constraint. WDYT?
https://github.com/longhorn/longhorn/issues/8332
cc @PhanLe1010
